packages/contracts/contracts/Router.sol

```diff
@@ -0,0 +1,220 @@
+import "@openzeppelin/contracts/access/Ownable.sol";
+import "@openzeppelin/contracts/utils/cryptography/ECDSA.sol";
+
+contract Router is Ownable {
+  address public immutable routerFactory;
+
+  ITransactionManager public transactionManager;
+
+  uint256 private chainId;
+
+  address public recipient;
+
+  address public routerSigner;
+
+  struct SignedPrepareData {
+    ITransactionManager.PrepareArgs args;
+    address routerRelayerFeeAsset;
+    uint256 routerRelayerFee;
+    uint256 chainId; // For domain separation
+  }
+
+  struct SignedFulfillData {
+    ITransactionManager.FulfillArgs args;
+    address routerRelayerFeeAsset;
+    uint256 routerRelayerFee;
+    uint256 chainId; // For domain separation
+  }
+
+  struct SignedCancelData {
+    ITransactionManager.CancelArgs args;
+    address routerRelayerFeeAsset;
+    uint256 routerRelayerFee;
+    uint256 chainId; // For domain separation
+  }
+
+  struct SignedRemoveLiquidityData {
+    uint256 amount;
+    address assetId;
+    uint256 chainId; // For domain separation
+  }
+
+  event RelayerFeeAdded(address assetId, uint256 amount, address caller);
+  event RelayerFeeRemoved(address assetId, uint256 amount, address caller);
+
+  constructor(address _routerFactory) {
+    routerFactory = _routerFactory;
+  }
+
+  // Prevents from calling methods other than routerFactory contract
+  modifier onlyViaFactory() {
+    require(msg.sender == routerFactory, "ONLY_VIA_FACTORY");
+    _;
+  }
+
+  function init(
+    address _transactionManager,
+    uint256 _chainId,
+    address _routerSigner,
+    address _recipient,
+    address _owner
+  ) external onlyViaFactory {
+    transactionManager = ITransactionManager(_transactionManager);
+    chainId = _chainId;
+    routerSigner = _routerSigner;
+    recipient = _recipient;
+    transferOwnership(_owner);
+  }
+
+  function setRecipient(address _recipient) external onlyOwner {
+    recipient = _recipient;
+  }
+
+  function setSigner(address _routerSigner) external onlyOwner {
+    routerSigner = _routerSigner;
+  }
+
+  function addRelayerFee(uint256 amount, address assetId) external payable {
+    // Sanity check: nonzero amounts
+    require(amount > 0, "#RC_ARF:002");
+
+    // Transfer funds to contract
+    // Validate correct amounts are transferred
+    if (LibAsset.isNativeAsset(assetId)) {
+      require(msg.value == amount, "#RC_ARF:005");
+    } else {
+      require(msg.value == 0, "#RC_ARF:006");
+      LibAsset.transferFromERC20(assetId, msg.sender, address(this), amount);
+    }
+
+    // Emit event
+    emit RelayerFeeAdded(assetId, amount, msg.sender);
+  }
+
+  function removeRelayerFee(uint256 amount, address assetId) external onlyOwner {
+    // Sanity check: nonzero amounts
+    require(amount > 0, "#RC_RRF:002");
+
+    // Transfer funds from contract
+    LibAsset.transferAsset(assetId, payable(recipient), amount);
+
+    // Emit event
+    emit RelayerFeeRemoved(assetId, amount, msg.sender);
+  }
```

Comment:
If someone compromises the owner, they could easily call `removeLiquidity` then `removeRelayerFee`. I don't think there's anyway around this, but it's important we spell that out somewhere in the documentation of the contracts

Reply:
The way "around it" is to have a sep signer as the remover/recipient of the liquidity, which it seems like we have. Enables routers to at least optionally have a diff signer that they keep even more securely (e.g. don't use it on their router's server, etc).

Reply:
if they compromise the owner key though, they can just remove all the relayer fee and liquidity but yeah that's just an effect of permissioning

Reply:
Making the remove recipient changeable via owner defeats the whole purpose of restricting the remove process in the first place.

Our current situation:
1. Remove liquidity can be called only by router (safe)
2. Router key is stored via web3signer (safe)
3. We need to expose some way to allow the router to call remove liquidity so we create an admin key/endpoint for it. This endpoint can send liquidity *anywhere* (unsafe)

With `recipient` changeable by owner:
1. Remove liquidity only callable by router key (safe)
2. Router key stored via web3signer (safe)
3. Admin key endpoint exists but funds can only go to address specified as part of setup (safe)
4. Deployer key which controls changing recipient now needs to be stored securely and recipient-changing functions now need to be exposed by admin endpoint (unsafe).

In particular, the above case creates an additional security assumption for router operators who are using a custody solution and expect to enforce that transactions go from custody solution to contract and back *without* any additional need tor keys/security beyond the existing custody system.

Instead of the above, we should explicitly *not* allow changing the recipient once contract is deployed. If routers need to change the recipient, they must then withdraw to the prespecified recipient address, and deploy a new wrapper contract, thus delegating the security entirely to the `recipient` (which is typically going to be a custody system).

Reply:
P.s. doing it this way more closely emulates the UX for DeFi liquidity provision which is the most expected user flow and has a lot more tooling build around it